src/content/blog/rust-based-javascript-linters-lack-typed-linting-right-now/index.mdx

```diff
@@ -0,0 +1,151 @@
+
+I think a more viable long-term option would be to find a way to get TypeScript's type checker to run at native speed.
+There are a couple possibilities:
+
+- Writing a tool that transpiles its source to a faster language such as Go or Rust
+- Pre-compiling and optimizing TypeScript like a binary
+
+Both of those options are difficult and will take some time to land.
+[Transpiling the checker to Go](https://kdy1.dev/2022-1-26-porting-tsc-to-go) was the original aim of what became `stc` before the [the project switched to the Rust re-implementation](https://kdy1.dev/2022-10-27-open-sourcing-stc#heading-ill-switch-back-to-the-rust-version).
+
+[Node.js user land snapshots](https://nodejs.org/en/blog/release/v18.8.0#bootstrap-implement-run-time-user-land-snapshots-via---build-snapshot-and---snapshot-blob) are mentioned in TypeScript's [Ideas for faster cold compiler start-up issue](https://github.com/microsoft/TypeScript/issues/25658) in the context of startup times.
+For the context of typed linting, aggressively optimizing code ahead of time might be marginally useful too.
+The [Hermes engine](https://hermesengine.dev) has some interesting build-time precompilation too.
+
+Though, TypeScript itself isn't architected for native code.
+It assumes a runtime with built-in garbage collection, mutable objects, and other performance paper cuts
+I suspect the biggest gains might be from rearchitecting TypeScript to be more performance-friendly:
+
+- Restricting global type augmentations to be more parallelization-friendly
+- Changing the way its checker runs to avoid those paper cuts
+
+Both avenues would be very difficult to implement and cause breaking changes in TypeScript's APIs.
+Neither is likely to happen any time soon.
+
+## Performance Comparisons
+
+I'm not going to show you a performance comparison of Rust-based linters vs. ESLint vs. ESLint with typescript-eslint.
+The comparison would be misleading: until Rust-based linters achieve feature parity with typed linting rules, they benefit in comparisons from having to run significantly less work.
+And given how many different avenues we have yet to flesh out in running type linting rules with a native speed linter, we have near-zero idea what that performance would look like.
+
+> 💡 When evaluating performance comparisons, always make sure the comparisons are on comparable behavior.
+> Don't trust any metric you don't understand the contents of.
+
+## In Conclusion
+
+Rust-based JavaScript/TypeScript linters such as [Biome](https://biomejs.dev), [`deno lint`](https://lint.deno.land), [oxc](https://github.com/oxc-project/oxc), and [RSLint](https://rslint.org) are fantastically fast projects.
+But that speed comes with a serious feature gap compared to [ESLint](https://eslint.org) + [typescript-eslint](https://typescript-eslint.io)'s [type-checked lint rules](https://typescript-eslint.io/linting/typed-linting).
+You should understand what you're missing when you switch from one to the other.
```

Comment:
The announcement I made for oxlint https://oxc-project.github.io/blog/2023-12-12-announcing-oxlint.html states the following:

> At this stage, oxlint is not intended to fully replace ESLint; it serves as an enhancement when ESLint's slowness becomes a bottleneck in your workflow.

> For faster feedback loops, we recommend running oxlint before ESLint in your lint-staged or CI setup, considering it only takes a few seconds to run on large codebases.

It would be appreciated if we emphasize this somehow, because Oxlint never advocated for replacement.

Reply:
Yeah! The Biome folks mentioned this too. I'll make it more clear+emphasized. Thanks!